enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,603 @@
+evident on GCP, as without it: the network plugin won't be able to perform any
+assignments at all, given the nodes' default `/32` subnet.
+
+#### Inter-component communication
+
+Multiple pieces of information needs to be communicated between the network
+plugins and this cloud component:
+
+1. The cloud's node subnet information.
+2. The egress IP assignment to be performed by the cloud component
+3. The status of the egress IP assignment performed by the cloud component
+
+Two models have been decided for this: a CRD based approach and a node
+annotated approach.
+
+##### annotated approach
+
+As to satisfy condition 1. the cloud component will annotate all cluster nodes
+with the nodes' cloud subnet as retrieved from the cloud provider's API when it
+is initialized. This annotation will have the following specification:
+
+`cloud.openshift.io/node-ifaddr: {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}`
+
+This will in turn be read by the network plugin and used during the egress IP
+assignment, superseding any host level subnet information already parsed by the
+network plugin.
+
+##### CRD
+
+The cloud component will define a dedicated CRD that the network plugin will use
+to issue cloud egress IP requests. The network plugin will in-turn finalize its
+setup and set its final egress IP status pending the CR's status indicating a
+successful/unsuccessful cloud assignment. For example: if the cloud assignment
+should be unsuccessful; it will be up to the network plugin to notify the user
+of this, by setting its own status field for the egress IP assignment
+accordingly.
+
+The CRD will have `scope: Cluster` (since egress IP is cluster-scoped and node
+private IP assignment have no relation to namespaces) and have the following
+format:
+
+```yaml
+apiVersion: cloud.openshift.io/v1
+kind: CloudPrivateIPConfig
+metadata:
+  name: cloud-egressip
+spec:
+  items:
+  - node: nodeX
+    ip: 192.168.126.11
+  - node: nodeY
+    ip: 192.168.126.102
+status:
+  items:
+  - node: nodeX
+    ip: 192.168.126.11
+  - node: nodeY
+    ip: 192.168.126.102
+```
+
+where `nodeX`/`nodeY` will be cluster node names as retrieved by
+`nodes[*].metadata.name`. The referential order between `.spec.items` and
+`.status.items` will not be strict, meaning `.spec.items[0]` will not
+necessarily correspond to `.spec.items[0]`.
+
+In case an assignment is unsuccessful: the `.status.items[?]` corresponding to
+that `.spec.items[?]` will not be set. This is done as to explicitly notify the
+network plugin of which (if not all) requests have failed by still issuing an
+UPDATE on the object. No further information with regards to the reason as to
+why the cloud egress IP request failed will be noted in this CRD. The reason for
+this is that the network plugins themselves do not provide such a field in the
+CRDs they manage for egress IP assignment. Any intention on doing so here, where
```

Comment:
"The downstream consumer is lame" is not an argument for making this API lame. :slightly_smiling_face: 

Especially, consider the case where the network plugin requests a single egress IP, and then the cloud IP provider doesn't update the status. Does that mean that it just hasn't processed the request yet, or that it tried and failed?

Reply:
> and then the cloud IP provider doesn't update the status.

I was actually thinking that the cloud component would do a "gratuitous empty update" which would trigger an update on the network plugin's side, only to notify the network plugin that nothing had been done. 

But sure, the consensus gravitates towards having a status field, so I will update the enhancement proposal with that. For completion's sake, I am proposing an enumerated and text based status field for each item. I.e:

```
apiVersion: cloud.openshift.io/v1
kind: CloudPrivateIPConfig
metadata:
  name: cluster
spec:
  items:
  - node: nodeX
    ip: 192.168.126.11
  - node: nodeY
    ip: 192.168.126.102
  - node: nodeZ
    ip: 192.168.126.103
status:
  items:
  - node: nodeX
    ip: 192.168.126.11
    status: SUCCESS 
    reason: ""
  - node: nodeY
    ip: 192.168.126.102
    status: PENDING
    reason: ""
  - node: nodeZ
    ip: 192.168.126.103
    status: ERROR
    reason: "Some reason why this would have failed"
```